src/scenic/syntax/veneer.py

```diff
@@ -111,1000 +132,1341 @@
-						  lambda self, dx, dy: Vector(dx, -self.length / 2 - dy))
-
-def leftSpecHelper(syntax, pos, dist, axis, toComponents, makeOffset):
-	extras = set()
-	if canCoerce(dist, builtins.float):
-		dx, dy = toComponents(coerce(dist, builtins.float))
-	elif canCoerce(dist, Vector):
-		dx, dy = coerce(dist, Vector)
-	else:
-		raise RuntimeParseError(f'"{syntax} X by D" with D not a number or vector')
-	if isinstance(pos, OrientedPoint):		# TODO too strict?
-		val = lambda self: pos.relativize(makeOffset(self, dx, dy))
-		new = DelayedArgument({axis}, val)
-		extras.add('heading')
-	else:
-		pos = toVector(pos, f'specifier "{syntax} X" with X not a vector')
-		val = lambda self: pos.offsetRotated(self.heading, makeOffset(self, dx, dy))
-		new = DelayedArgument({axis, 'heading'}, val)
-	return Specifier('position', new, optionals=extras)
+    """The :specifier:`behind {X} by {Y}` polymorphic specifier.
+
+    Specifies :prop:`position`, depending on :prop:`length`.
+
+    Allowed forms::
+
+        behind <oriented point> [by <scalar/vector>]
+        behind <vector> [by <scalar/vector>]
+
+    If the :grammar:`by <scalar/vector>` is omitted, zero is used.
+    """
+    return directionalSpecHelper('Behind', pos, dist, 'length', lambda dist: (0, dist, 0),
+                          lambda self, dims, tol, dx, dy, dz: Vector(dx, -self.length / 2 - dy - dims[1]/2 - tol, dz))
+
+def Above(pos, dist=0):
+    """The :specifier:`above {X} by {Y}` polymorphic specifier.
+
+    Specifies :prop:`position`, depending on :prop:`height`. 
+
+    Allowed forms::
+
+        above <oriented point> [by <scalar/vector>]
+        above <vector> [by <scalar/vector>]
+
+    If the :grammar:`by <scalar/vector>` is omitted, zero is used.
+    """
+    return directionalSpecHelper('Above', pos, dist, 'height', lambda dist: (0, 0, dist),
+                          lambda self, dims, tol, dx, dy, dz: Vector(dx, dy, self.height / 2 + dz + dims[2]/2 + tol))
+
+def Below(pos, dist=0):
+    """The :specifier:`below {X} by {Y}` polymorphic specifier.
+
+    Specifies :prop`position`, depending on :prop:`height`.
+
+    Allowed forms::
+
+        below <oriented point> [by <scalar/vector>]
+        below <vector> [by <scalar/vector>]
+
+    If the 'by <scalar/vector>' is omitted, zero is used.
+    """
+    return directionalSpecHelper('Below', pos, dist, 'height', lambda dist: (0, 0, dist),
+                          lambda self, dims, tol, dx, dy, dz: Vector(dx, dy, -self.height / 2 - dz - dims[2]/2 - tol))
+
+def directionalSpecHelper(syntax, pos, dist, axis, toComponents, makeOffset):
+    prop = {'position': 1}
+    if canCoerce(dist, builtins.float):
+        dx, dy, dz = toComponents(coerce(dist, builtins.float))
+    elif canCoerce(dist, Vector):
+        dx, dy, dz = coerce(dist, Vector)
+    else:
+        raise TypeError(f'"{syntax} X by D" with D not a number or vector')
+
+    @distributionFunction
+    def makeContactTol(dist, ct):
+        if dist == 0:
```

Comment:
The documentation says the contact tolerance is used when no distance is given; this code also uses it when the distance is given as 0. I think the former makes more sense: if somebody wrote `left of X by Y` where `Y` is an expression it would be strange for the position to jump discontinuously as `Y` decreases to zero.